File: channels-nk/trunk/mod-data/modules/mod-data/libs/mutex.js
// Implements http://www.eonclash.com/Tutorials/Multithreading/MartinHarvey1.1/Ch11.html
/*
    * Read operations can execute concurrently.
    * Write operations cannot execute at the same time as read operations.
    * Write operations cannot execute at the same time as write operations.
    * 
    * There is an asymmetry in the synchronization scheme: threads potentially wanting to read 
    * will block before reading if there are any active writers, whilst threads wanting to write 
    * block before writing if there are any reading readers. This gives priority to writing threads; 
    * a sensible approach, given that writes are less frequent than reads.
 */
 
LOG_MUTEX = true;

function issueMutexRequest(command, who) {
	var req = context.createSubRequest("active:MREWSynchronizer");
	req.addArgument("operator", new StringAspect(command));
	if (LOG_MUTEX) log(who + ": Start " + command, "info");
	context.issueSubRequest(req);
	if (LOG_MUTEX) log(who + ": End " + command, "info");
}

// Multiple reads, exclusive write

function beginRead(who) {
	issueMutexRequest("beginRead", who);
}

function endRead(who) {
	issueMutexRequest("endRead", who);
}

function beginWrite(who) {
	issueMutexRequest("beginWrite", who);
}

function endWrite(who) {
	issueMutexRequest("endWrite", who);
}

File: channels-nk/trunk/mod-data/modules/mod-data/com/mindalliance/channels/data/Mutex.java
// Copyright (C) 2007 Mind-Alliance Systems LLC.
// All rights reserved.

package com.mindalliance.channels.data;

import org.ten60.netkernel.layer1.nkf.INKFConvenienceHelper;
import org.ten60.netkernel.layer1.nkf.INKFRequest;
import org.ten60.netkernel.layer1.nkf.NKFException;
import org.ten60.netkernel.layer1.representation.StringAspect;

import com.mindalliance.channels.nk.ContextHelper;


public class Mutex {

    private boolean LOG_MUTEX = true;
    private ContextHelper contextHelper;
    
    public Mutex(INKFConvenienceHelper context) {
        contextHelper = new ContextHelper(context);
    }

    private void issueMutexRequest(String command, String who) throws NKFException {
        INKFRequest req = contextHelper.context.createSubRequest("active:MREWSynchronizer");
        req.addArgument("operator", new StringAspect(command));
        if (LOG_MUTEX) contextHelper.log(who + ": Start " + command, "info");
        contextHelper.context.issueSubRequest(req);
        if (LOG_MUTEX) contextHelper.log(who + ": End " + command, "info");
    }

//     Multiple reads, exclusive write

    public void beginRead(String who) throws NKFException {
        issueMutexRequest("beginRead", who);
    }

    public void endRead(String who) throws NKFException {
        issueMutexRequest("endRead", who);
    }

    public void beginWrite(String who) throws NKFException {
        issueMutexRequest("beginWrite", who);
    }

    public void endWrite(String who) throws NKFException {
        issueMutexRequest("endWrite", who);
    }
    
}
